### PR TITLE
Expose all rewards (fees, rent, voting and staking) in RPC getConfirmedBlock and the cli

### PR DESCRIPTION
**Pull request by mvines**
The rewards array in `getConfirmedBlock` has a couple limitations that this PR addresses:
1. No indication of the type of reward -- was it for voting and staking?  This can be inferred by looking at the type of account of course but that's clumsy
2. Transaction fees and rent distributions are not surfaced

With this PR `getConfirmedBlock` will now report at minimum one reward per block, the transaction fees paid to the leader.

### CODE REVIEW

File: cli/src/cluster_query.rs [hunk 1]
```diff
@@ -73,8 +73,7 @@ impl ClusterQuerySubCommands for App<'_, '_> {
                         .validator(is_slot)
```
Comment (mvines) [inline cli/src/cluster_query.rs:73]: `solana block` now dumps the latest confirmed block instead an error, which is a nice feature when watching for transaction fees.

File: core/src/replay_stage.rs [hunk 2]
```diff
@@ -291,7 +291,6 @@ impl ReplayStage {
                         &bank_forks,
```
Comment (mvines) [inline core/src/replay_stage.rs:291]: ReplayStage needed some rejigging.  `record_rewards` needs to happen after the bank is frozen instead of immediately upon Bank creation

File: storage-bigtable/src/convert.rs [hunk 3]
```diff
@@ -24,6 +25,13 @@ impl From<Reward> for generated::Reward {
             pubkey: reward.pubkey,
```
Comment (mvines) [inline storage-bigtable/src/convert.rs:25]: Getting enums in/out of protobuf is a little clunky

File: transaction-status/src/lib.rs [hunk 4]
```diff
@@ -241,6 +242,8 @@ pub struct Reward {
     pub lamports: i64,
     #[serde(deserialize_with = "default_on_eof")]
     pub post_balance: u64, // Account balance in lamports after `lamports` was applied
+    #[serde(default, deserialize_with = "default_on_eof")]
```
Comment (mvines) [inline transaction-status/src/lib.rs:245]: `#[serde(deserialize_with = "default_on_eof")]` was insufficient to satisfy `serde_json` when trying to fetch an existing confirmed block from testnet without the `reward_type` field.  Adding `#[serde(default)]` as well made it happy

File: storage-bigtable/src/convert.rs (hunk 3 above)
Comment (mvines) [reply storage-bigtable/src/convert.rs:25]: That doesn't seem much better, I'll keep the existing code I think

File: transaction-status/src/lib.rs (hunk 4 above)
Comment (mvines) [reply transaction-status/src/lib.rs:245]: With this PR, remove the `default` and then:
```
$ cargo run -- --verbose block 40700256
    Finished dev [unoptimized + debuginfo] target(s) in 0.34s
     Running `/Users/mvines/ws/s2/target/debug/solana --verbose block 40700256`
RPC URL: http://testnet.solana.com
Default Signer Path: /Users/mvines/.config/solana/id.json
Error: missing field `rewardType`
```

serde_json doesn't even hit the `default_on_eof` deserializer, it already decided that there's a missing field in the response and it can't do anything about it.  Deriving `Default` for the entire `Reward` struct produces the same error

File: runtime/src/bank.rs [hunk 5]
```diff
@@ -523,8 +523,32 @@ impl PartialEq for Bank {
     }
 }
 
-#[derive(Debug, PartialEq, Serialize, Deserialize, AbiExample, Default, Clone, Copy)]
+#[derive(Debug, PartialEq, Serialize, Deserialize, AbiExample, AbiEnumVisitor, Clone, Copy)]
+pub enum RewardType {
+    Fee,
+    Rent,
+    Voting,
+    Staking,
```
[the remark this answers is not in the document]
Comment (mvines) [reply runtime/src/bank.rs:531]: Sure, done.  Of course the next time we add a reward, it needs to start W, X, Y or Z in order to keep things alphabetical

File: runtime/src/bank.rs [hunk 6]
```diff
@@ -5717,14 +5791,15 @@ mod tests {
 
         // verify validator rewards show up in bank1.rewards vector
         assert_eq!(
-            bank1.rewards,
-            Some(vec![(
+            *bank1.rewards.read().unwrap(),
+            vec![(
                 stake_id,
                 RewardInfo {
+                    reward_type: RewardType::Staking,
```
[the remark this answers is not in the document]
Comment (mvines) [reply runtime/src/bank.rs:5863]: I snuck a little something into `test_bank_update_rewards_determinism()`

File: transaction-status/src/lib.rs (hunk 4 above)
Comment (CriesofCarrots) [reply transaction-status/src/lib.rs:245]: Yeah, `serde_json` doesn't need the `default_on_eof`. That's for bincode going into and out of blockstore.

...Ah, okay, figured it out. I was curious why this isn't affection `log_messages`, which also aren't on testnet right now. It's because the meta are being parsed by serde_json into `UiTransactionStatusMeta`, which don't include the `default_on_eof` tag.
Similarly, the serde_json parsing of `Reward` from rpc works fine with the `default_on_eof` tag removed. 

tl;dr, we'll need to remember `#[serde(default)]` for new `default_on_eof` fields on structs that are used directly in rpc.

File: runtime/src/bank.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply runtime/src/bank.rs]: Challenge accepted! 😉

**Review by CriesofCarrots — APPROVED** [review]
